projects/npm-tools/packages/npm-scripts/test/typescript/getTypeScriptBuildOrder.js

```diff
@@ -0,0 +1,234 @@
+		expect(getTypeScriptBuildOrder({a, d, b, c})).toEqual([c, a, d, b]);
+		expect(getTypeScriptBuildOrder({a, d, c, b})).toEqual([c, a, d, b]);
+		expect(getTypeScriptBuildOrder({b, a, c, d})).toEqual([c, a, d, b]);
+		expect(getTypeScriptBuildOrder({b, a, d, c})).toEqual([c, a, d, b]);
+		expect(getTypeScriptBuildOrder({b, c, a, d})).toEqual([c, a, d, b]);
+		expect(getTypeScriptBuildOrder({b, c, d, a})).toEqual([c, a, d, b]);
+		expect(getTypeScriptBuildOrder({b, d, a, c})).toEqual([c, a, d, b]);
+		expect(getTypeScriptBuildOrder({b, d, c, a})).toEqual([c, a, d, b]);
+		expect(getTypeScriptBuildOrder({c, a, b, d})).toEqual([c, a, d, b]);
+		expect(getTypeScriptBuildOrder({c, a, d, b})).toEqual([c, a, d, b]);
+		expect(getTypeScriptBuildOrder({c, b, d, a})).toEqual([c, a, d, b]);
+		expect(getTypeScriptBuildOrder({c, b, d, a})).toEqual([c, a, d, b]);
+
+		// But note that we can get different (but still valid) orderings:
+
+		expect(getTypeScriptBuildOrder({c, d, a, b})).toEqual([c, d, a, b]);
+		expect(getTypeScriptBuildOrder({c, d, b, a})).toEqual([c, d, a, b]);
+		expect(getTypeScriptBuildOrder({d, a, b, c})).toEqual([c, d, a, b]);
+		expect(getTypeScriptBuildOrder({d, a, c, b})).toEqual([c, d, a, b]);
+		expect(getTypeScriptBuildOrder({d, b, a, c})).toEqual([c, d, a, b]);
+		expect(getTypeScriptBuildOrder({d, b, c, a})).toEqual([c, d, a, b]);
+		expect(getTypeScriptBuildOrder({d, c, a, b})).toEqual([c, d, a, b]);
+		expect(getTypeScriptBuildOrder({d, c, b, a})).toEqual([c, d, a, b]);
+
+		// And we can get a similar effect by changing the order of
+		// dependencies _inside_ "b"; again, a different (but still
+		// valid) ordering.
+
+		const b2 = {name: 'b2', dependencies: {d: '*', a: '*'}};
+
+		expect(getTypeScriptBuildOrder({b2, a, c, d})).toEqual([c, d, a, b2]);
+
+		/* eslint-enable sort-keys */
+	});
+
+	it('handles a more complex example', () => {
+
+		//            Y
+		//            | \
+		//        Z   X  \   ^
+		//       /|\ /|   \  | dependencies point upwards
+		//      J K L M N |  |
+		//      | | |\|/ /
+		//      A B C D-/
+
+		/* eslint-disable sort-keys */
+
+		const a = {name: 'a', dependencies: {j: '*'}};
+		const b = {name: 'b', dependencies: {k: '*'}};
+		const c = {name: 'c', dependencies: {l: '*'}};
+		const d = {name: 'd', dependencies: {l: '*', m: '*', n: '*', y: '*'}};
+		const j = {name: 'j', dependencies: {z: '*'}};
+		const k = {name: 'k', dependencies: {z: '*'}};
+		const l = {name: 'l', dependencies: {z: '*', x: '*'}};
+		const m = {name: 'm', dependencies: {x: '*'}};
+		const n = {name: 'n', dependencies: {}};
+		const x = {name: 'x', dependencies: {y: '*'}};
+		const y = {name: 'y', dependencies: {}};
+		const z = {name: 'z', dependencies: {}};
+
+		expect(
+			getTypeScriptBuildOrder({a, b, c, d, j, k, l, m, n, x, y, z})
+		).toEqual([z, j, a, k, b, y, x, l, c, m, n, d]);
+
+		/* eslint-enable sort-keys */
+	});
+
+	it('complains if it detects a cycle', () => {
```

Comment:
Is this a hard limitation? I've briefly googled for TS cyclic refs and there seems to be some literature but didn't fully get into it. I know deps loading order is solvable to a certain extent in the JS context, so wondering how's that different here (I assume partial type loading/compilation is not a thing)

What should we tell a developer that faces this issue? How can we provide some guidance on how to refactor their code to avoid cycles?

Reply:
AFAIK, it is a hard limitation. You cannot do a topological sort of a dependency graph that has cycles because, by definition, there _is_ no order in which you can guarantee that every item's dependencies have been built before the item itself. It is possible that `tsc` might have some magic in there to support cycles between projects (now or in the future); I didn't even search, beyond noting that if you try building `frontend-js-react-web` before `frontend-js-state-web` has produced its artifacts, you get an error:

```
error TS6305: Output file
'$PORTAL/modules/apps/frontend-js/frontend-js-state-web/build/node/packageRunBuild/dist/src/main/resources/META-INF/resources/index.d.ts'
has not been built from source file
'$PORTAL/modules/apps/frontend-js/frontend-js-state-web/src/main/resources/META-INF/resources/index.ts'.
```

There are some forms of cyclic refs that are fine in TS, but those won't be project-level refs. For example, given two files in the same project: a file `a.ts` can import a _type_ from `b.ts`, and `b.ts` can import stuff (types or values) from `a.ts`; in fact, circular type imports aren't even uncommon in TS in my experience. But circular _value_ imports should be avoided as a matter of hygiene, IMO.

Our bundler/loader didn't support circular imports for a while, IIRC, and for good reason. I think we ended up hacking support in there only because there were misbehaved third-party deps that had cycles in them, and we couldn't control those.

As far as advising devs on how to avoid cycles, the answer is always the same: hoist — if `foo` and `bar` depend mutually on one another, extract cyclic part up into `common` so that the mutually dependent bits live in the same file. In some cases, this will actually eliminate `foo` and/or `bar`, but in others there may be some remnant; either way though, the dependency graph becomes acyclic. If the cycle is longer (eg. `foo` depends on `bar` depends on `baz` depends on `foo`), the answer is still to hoist.

Reply:
Do you have any concrete examples in mind? I'm aware of `frontend-js-web` + `frontend-js-aui-web` having a mysterious spooky-at-a-distance mutual dependency because they cohabit the same space via the `FragmentHost` mechanism, and parts of `Liferay` are actually set up by JSP etc. But I am not too worried about `aui-web` as we shouldn't really be touching it (other than to delete it). Any other instances to have on our radar?

Reply:
> Do you have any concrete examples in mind? 

I think `frontend-taglib-clay` <-> `frontend-js-react-web` (with or without `frontend-js-web` in between). We know the react one uses `ClayIconSpriteContext` when rendering and that in the clay one we likely use hooks offered by the react module.

Reply:
Taglibs, you say...

![taglibs](https://user-images.githubusercontent.com/7074/113164139-22d75900-9241-11eb-966f-fff2e892b987.gif)

Reply:
![smellofnewmodules](https://user-images.githubusercontent.com/905006/113164870-6a59e700-9239-11eb-8a0f-18f543e8bed5.jpeg)